components/eamxx/src/diagnostics/entrainment_budget.cpp

```diff
@@ -0,0 +1,300 @@
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<>(0, ncols * nlevs), KOKKOS_LAMBDA(const int &idx) {
+        const int icol   = idx / ncols;
+        const int jlev   = idx % nlevs;
+        qt_v(icol, jlev) = qc_v(icol, jlev) + qv_v(icol, jlev);
+        tl_v(icol, jlev) = PF::calculate_thetal_from_theta(
+            PF::calculate_theta_from_T(tm_v(icol, jlev), pm_v(icol, jlev)),
+            tm_v(icol, jlev), qc_v(icol, jlev));
+      });
+}
+
+void EntrainmentBudget::init_timestep(const util::TimeStamp &start_of_step) {
+  m_start_t = start_of_step;
+
+  const auto &tm_v = get_field_in("T_mid").get_view<Real **>();
+  const auto &pm_v = get_field_in("p_mid").get_view<Real **>();
+  const auto &qv_v = get_field_in("qv").get_view<Real **>();
+  const auto &qc_v = get_field_in("qc").get_view<Real **>();
+
+  const auto &m_prev_qt_v = m_prev_qt.get_view<Real **>();
+  const auto &m_prev_tl_v = m_prev_tl.get_view<Real **>();
+
+  calc_tl_qt(tm_v, pm_v, qv_v, qc_v, m_prev_tl_v, m_prev_qt_v);
+}
+
+void EntrainmentBudget::compute_diagnostic_impl() {
+  using PC  = scream::physics::Constants<Real>;
+  using KT  = KokkosTypes<DefaultDevice>;
+  using MT  = typename KT::MemberType;
+  using ESU = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+
+  constexpr Real g = PC::gravit;
+
+  // Before doing anything, subview the out field for each variable
+  auto out = m_diagnostic_output.get_view<Real **>();
+
+  auto o_pm_hplus = ekat::subview_1(out, m_index_map["p+"]);
+  auto o_tl_hplus = ekat::subview_1(out, m_index_map["tl+"]);
+  auto o_tl_caret = ekat::subview_1(out, m_index_map["tl^"]);
+  auto o_tl_ttend = ekat::subview_1(out, m_index_map["tl_ttend"]);
+  auto o_qt_hplus = ekat::subview_1(out, m_index_map["qt+"]);
+  auto o_qt_caret = ekat::subview_1(out, m_index_map["qt^"]);
+  auto o_qt_ttend = ekat::subview_1(out, m_index_map["qt_ttend"]);
+  auto o_df_inpbl = ekat::subview_1(out, m_index_map["dF"]);
+
+  // Get the input views
+  const auto &qc_v = get_field_in("qc").get_view<Real **>();
+  const auto &qv_v = get_field_in("qv").get_view<Real **>();
+  const auto &tm_v = get_field_in("T_mid").get_view<Real **>();
+  const auto &pm_v = get_field_in("p_mid").get_view<Real **>();
+  const auto &pd_v = get_field_in("pseudo_density").get_view<Real **>();
+  const auto &sd_v = get_field_in("SW_flux_dn").get_view<Real **>();
+  const auto &su_v = get_field_in("SW_flux_dn").get_view<Real **>();
+  const auto &ld_v = get_field_in("LW_flux_dn").get_view<Real **>();
+  const auto &lu_v = get_field_in("LW_flux_up").get_view<Real **>();
+
+  // tracked stuff
+  const auto &prev_qtot_v = m_prev_qt.get_view<Real **>();
+  const auto &prev_tliq_v = m_prev_tl.get_view<Real **>();
+
+  view_2d qt_v("qt_v", m_ncols, m_nlevs);
+  view_2d tl_v("tl_v", m_ncols, m_nlevs);
+  calc_tl_qt(tm_v, pm_v, qv_v, qc_v, tl_v, qt_v);
+
+  const auto &curr_ts =
+      get_field_in("qc").get_header().get_tracking().get_time_stamp();
+
+  std::int64_t dt = curr_ts - m_start_t;
```

Comment:
I'd use `auto` here.

Reply:
Yes. The return type of op- may or may not change in the future, so unless you _need_ to use a special type, auto is best.